gdal/swig/python/scripts/gdal_calc.py

```diff
@@ -190,61 +274,97 @@ def doit(opts, args):
+
         myOut = gdal.Open(opts.outF, gdal.GA_Update)
-        if [myOut.RasterXSize, myOut.RasterYSize] != DimensionsCheck:
-            raise Exception("Error! Output exists, but is the wrong size.  Use the --overwrite option to automatically overwrite the existing file")
+        if myOut is None:
+            error = 'but cannot be opened for update'
+        elif [myOut.RasterXSize, myOut.RasterYSize] != DimensionsCheck:
+            error = 'but is the wrong size'
+        elif ProjectionCheck and ProjectionCheck != myOut.GetProjection():
+            error = 'but is the wrong projection'
+        elif GeoTransformCheck and GeoTransformCheck != myOut.GetGeoTransform():
+            error = 'but is the wrong geotransform'
+        else:
+            error = None
+        if error:
+            raise Exception("Error! Output exists, %s.  Use the --overwrite option to automatically overwrite the existing file" % error)
+
         myOutB = myOut.GetRasterBand(1)
         myOutNDV = myOutB.GetNoDataValue()
-        myOutType = gdal.GetDataTypeName(myOutB.DataType)
+        myOutType = myOutB.DataType
 
     else:
-        # remove existing file and regenerate
-        if os.path.isfile(opts.outF):
-            os.remove(opts.outF)
-        # create a new file
-        if opts.debug:
-            print("Generating output file %s" % (opts.outF))
+        if opts.outF:
+            # remove existing file and regenerate
+            if os.path.isfile(opts.outF):
+                os.remove(opts.outF)
+            # create a new file
+            if opts.debug:
+                print("Generating output file %s" % (opts.outF))
+        else:
+            opts.outF = ''
 
         # find data type to use
         if not opts.type:
             # use the largest type of the input files
-            myOutType = gdal.GetDataTypeName(max(myDataTypeNum))
+            myOutType = max(myDataTypeNum)
         else:
             myOutType = opts.type
+            if isinstance(myOutType, str):
+                myOutType = gdal.GetDataTypeByName(myOutType)
 
         # create file
         myOutDrv = gdal.GetDriverByName(opts.format)
         myOut = myOutDrv.Create(
             opts.outF, DimensionsCheck[0], DimensionsCheck[1], allBandsCount,
-            gdal.GetDataTypeByName(myOutType), opts.creation_options)
+            myOutType, opts.creation_options)
 
         # set output geo info based on first input layer
-        myOut.SetGeoTransform(myFiles[0].GetGeoTransform())
-        myOut.SetProjection(myFiles[0].GetProjection())
+        if not GeoTransformCheck:
+            GeoTransformCheck = myFiles[0].GetGeoTransform()
+        myOut.SetGeoTransform(GeoTransformCheck)
+        if not ProjectionCheck:
+            ProjectionCheck = myFiles[0].GetProjection()
+        myOut.SetProjection(ProjectionCheck)
 
         if opts.NoDataValue is not None:
-            myOutNDV = opts.NoDataValue
+            if opts.NoDataValue is ...:
```

Comment:
pardon my Python ignorance, but what does "is ..."  mean ?

Reply:
Ellipsis (...) is a singleton object, like None. 
It has a few use cases. Here I just use it to say "default value".
https://stackoverflow.com/questions/54405918/python-3-ellipsis-in-function-parameters

If backwards compatibility was not an issue, I would have done it differently:
opts.NoDataValue=None to indicate that there is no noDataValue.
opts.NoDataValue=... to indicate a default no data value
but since opts.NoDataValue=None already maps to the default value, I needed to choose a different value for noDataValue that would translate to noDataValue, a value that is not a valid noDataValue - thus I chose ellipsis.

Reply:
Apparently ellipsis was added in Python3.5.
Do you have a better idea for indicating not to set a nodata value? 
maybe: 
if not opts.NoDataValue and opts.NoDataValue is not None
so an empty string, or an empty tuple will indicate that?

Reply:
I'm afraid I'm a bit lost (I'm only modestly familiar with this script). If the user has not defined NoDataValue, my understanding of the code was that at this point opts.NoDataValue should be None. For me either it is None or a numeric value. Can't think of anyone else.

Reply:
right.
it's either None, which indicates the default value as per datatype
or a numeric value which indicates the requested nodatavalue.
I want to add an option for not setting a nodata value at all.
which, for I would have done it by setting it to None, but it's already has a meaning of a default nodata value.
thus I thought to allow opts.NoDataValue=[] to indicate not to set nodatavalue at all.

Reply:
I've committed this change, I hope it makes scene.

Reply:
ah, I better understand now what you wanted to achieve. In other utilities we use the string 'none' (case insensitive. could be 'NONE') for such a situation ("gdal_translate -a_nodata none"). I'd suggest to use that convention too.

Reply:
OK, I used "if opts.NoDataValue in ((), [], '', 'none', 'NONE')" to cover all conventions...